core/templates/dev/head/pages/exploration_player/progress_nav_directive.html

```diff
@@ -1,12 +1,40 @@
-<div ng-if="hasNext() || hasPrevious()" layout="row" layout-align="space-between center" style="background-color: white;">
+<div layout="row" layout-align="space-between center" style="background-color: white;">
   <!-- Wrapping div on first button keeps the second button right-aligned
     even when the first button is not present. -->
   <div>
-    <md-button ng-if="hasPrevious()" ng-click="changeCard(currentCardIndex-1)" class="md-raised" style="margin: 0 0 12px 12px;">
+    <md-button ng-if="hasPrevious" ng-click="changeCard(activeCardIndex-1)" class="md-raised" style="margin: 0 0 12px 12px;">
       <i class="material-icons oppia-vcenter" aria-hidden="true">&#xE5C4;</i>
     </md-button>
   </div>
-  <md-button ng-if="hasNext()" ng-click="changeCard(currentCardIndex+1)" class="md-raised" style="margin: 0 12px 12px 0">
+  <md-button ng-if="hasNext" ng-click="changeCard(activeCardIndex+1)" class="md-raised" style="margin: 0 12px 12px 0">
     <i class="material-icons oppia-vcenter" aria-hidden="true">&#xE5C8;</i>
   </md-button>
+  <div ng-if="!hasNext" style="margin: 0 12px 12px 0">
+    <div ng-if="!showContinueButton() && !helpCardHasContinueButton">
+      <!-- TODO(allan): Look at using ng-switch once we upgrade
+        to >=angular-1.5.10, which supports "ng-switch-when-separator".
+        However, the current approach is cleaner in angular-1.5.8.
+      -->
+      <md-button ng-if="interaction.id === 'Continue'"
+        class="oppia-learner-confirm-button" ng-click="submitFn()">
+        <[interaction.customizationArgs.buttonText.value]>
+        <i class="fa fa-arrow-right" style="font-size: 19px; padding-top: 1.5px;"></i>
+      </md-button>
+      <md-button ng-if="interaction.id === 'TextInput' ||
```

Comment:
What about item selection input? Are there others?

I'm quite concerned about maintainability here. It would be better for each interaction's .py file to declare a includes_submit_button field instead of hard-coding here. Otherwise I don't see future interaction creators remembering to update this.

Reply:
@AllanYangZhou and I discussed this in chat. Here's a summary:

- Let's add a boolean property show_generic_submit_button_in_nav to the interaction .py files. Set it to True for the six interactions mentioned here and False for the others. (Also check parity with other interactions, e.g. ItemSelectionInput.)
  - Add a comment above that property in base.py that's something like "Whether to show a submit button in the navigation area when the window width is small. Developers should ensure that their interaction JS binds the XXX event, and hide the submit button from the main interaction when the window width is small". 
  - Also, add a comment above that property in the Continue button saying "The Continue button does appear in the nav, but we handle it separately."
  - Then, here, do ng-if="interaction.show_submit_button".
- There's also another issue, which is that an interaction developer may, in the future, modify the code interaction's submit button to "Run and Submit", or add a play icon to it, or something.
  - Further work needs to be done so that a generic button is inserted on the bottom right of the interaction automatically, and there is no dev-written code for that button in particular in all of the existing interactions -- or, alternatively, there should be some backend/frontend test that ensures the conditions specified always hold (maybe pull the standard button into its own directive and check that the interaction includes that code snippet exactly). @AllanYangZhou will file an issue for this with more details.
  - In the meantime, let's add a comment above each Submit button in those interactions, and also above the corresponding JS code -- something like "this should be a standard Submit button otherwise it'll break parity with the interaction's show_generic_submit_button_in_nav setting".

Reply:
Ah, one thing that didn't occur to me is that ItemSelectionInput is already an example of the 2nd bullet point, it shows a Submit button only in some cases (when maxAllowableSelectionCount > 1) .

So I'm not sure if we can put that off actually.

Reply:
Sigh. For now, let's special-case it like we do for the continue button. I can't think of a better way to do this that's still reasonable to implement in the short-term.